lib/client.js

```diff
@@ -1092,29 +1112,43 @@ async function connect (client) {
       return
     }
 
-    if (!llhttpInstance) {
-      llhttpInstance = await llhttpPromise
-      llhttpPromise = null
-    }
-
     client[kConnecting] = false
 
     assert(socket)
 
-    socket[kNoRef] = false
-    socket[kWriting] = false
-    socket[kReset] = false
-    socket[kBlocking] = false
-    socket[kError] = null
-    socket[kParser] = new Parser(client, socket, llhttpInstance)
-    socket[kClient] = client
-    socket[kCounter] = 0
-    socket[kMaxRequests] = client[kMaxRequests]
-    socket
-      .on('error', onSocketError)
-      .on('readable', onSocketReadable)
-      .on('end', onSocketEnd)
-      .on('close', onSocketClose)
+    if (socket.alpnProtocol === 'h2') {
+      const session = http2.connect(client[kUrl], {
+        createConnection: () => socket
+      })
+
+      session[kError] = null
+      session[kClient] = client
+      session.on('error', onHttp2SessionError)
+      session.on('close', onSocketClose)
+      session.unref()
+
+      client[kHTTP2Session] = session
```

Comment:
I think concurrency should be handled through an LRU cache scoped to the `Client` class. This approach would be very similar to @szmarczak  https://github.com/szmarczak/http2-wrapper library.

### Public Interface
First I think the addition of http2 should expose the `http2.connect` options to whomever is instantiating the `Client` class through an `http2` constructor option. Maybe even possibly omitting the `createConnection` option? 

### Performance
One key aspect of performant http2 is setting the appropriate window sizing. That being said the default `http2` window sizing should be resolved at some point. See https://github.com/nodejs/node/issues/38426

Until the default window sizing issue is resolved the initial window size should be set;
```js
const session = http2.connect(client[kUrl], {
  createConnection: () => socket,
  initialWindowSize: 1024 * 1024 * 32
})
```

### Concurrency Events
If the LRU cache for concurrency approach is taken, the `session` events for `remoteSettings` and `goaway` need to be handled.

### Remote Settings
Once the session receives a `remoteSettings` settings event, the session should set a`Client` class `kMaxConcurrentStreams` limit and set the local window size of the session.

```js
// See https://github.com/nodejs/node/issues/38426
if (session.setLocalWindowSize) {
  session.setLocalWindowSize(1024 * 1024 * 4); // 4 MB
}
```

Within the handler for once the session receives a `remoteSettings` settings event, another listener should be created for on `remoteSettings` as the session `session.remoteSettings.maxConcurrentStreams`  needs to update the value for the `kMaxConcurrentStreams`.

### Goaway
If concurrency is implemented through an LRU cache, once a `goaway` event is received from the remote the session should be removed from the LRU cache so it won't be reused.

Reply:
Ah, I see, the client instances are maintained in the `Pool` class. Thus the remote settings and http2 connection state/settings need to be bubbled up.

Reply:
Thanks for all the feedback, incredibly insightful 🙇 

Going one by one:

>First I think the addition of http2 should expose the http2.connect options to whomever is instantiating the Client class through an http2 constructor option. 

I would love to hear @mcollina, @ronag, and @KhafraDev thoughts to see which options would be ok to move to the public interface, and based on that expose the one required (at least the bare minium). It is also worth it to say that Undici already exposes configs required to manipulate the `Socket` connection

> Maybe even possibly omitting the createConnection option?

Not so sure about this one. By the nature of Undici, the initial connection is made through a socket assuming an H1 connection. If then, the server negotiates and serves a H2, is when this feature comes into play.


>One key aspect of performant http2 is setting the appropriate window sizing. That being said the default http2 window sizing should be resolved at some point. See https://github.com/nodejs/node/issues/38426

Good catch, seems is already approved. I'll update it then 👍 


**For parallel streams**
Though I agree with your points, I didn't start yet with the parallel streams supports, even tho is something that should be treated with care as you suggest.
But this requires also come with a good default, in combination with an API to update several settings accordingly for concurrency.
I would like to hear thoughts from the team, in the meanwhile I'll implement the smaller pieces already pointed out and kick-off benchmarks as well 😄